File: scripts/datapull.py
#!/usr/bin/env python
import json
import os
import urllib2

from datetime import datetime
from pprint import pprint

import pymongo
from pymongo.errors import (DuplicateKeyError,
                            OperationFailure)

uri = os.environ.get('MONGOLAB_URI')

c = pymongo.Connection(uri)
db = c.githubdata
coll = db.repos

# github's paging seems to return dupes...
coll.ensure_index([('username', pymongo.ASCENDING),
                   ('name', pymongo.ASCENDING)], unique=True)

fmt = "%Y-%m-%dT%H:%M:%S"

page = 0

while True:

    #print(page)
    response = urllib2.urlopen('https://api.github.com/legacy/'
        'repos/search/:mongodb?start_page'
        '=%d' % (page,))

    data = response.read()
    js = json.loads(data)
    # github returns an empty array for 'repositories'
    # when there are no more results.
    if js['repositories']:
        for repo in js['repositories']:
            # Convert date strings to real timestamps.
            for field in ('created', 'created_at', 'pushed', 'pushed_at'):
                # TODO: This sucks ass but skip timezone offsets for the time being.
                repo[field] = datetime.strptime(repo[field][:-6], fmt)
            try:
                # Update so we get updated followers and watchers count.
                res = coll.update({'username': repo['username'],
                                   'name': repo['name']},
                                  repo, upsert=True, safe=True)
                if not res['updatedExisting']:
                    pass
            except (DuplicateKeyError, OperationFailure):
                pass

        page += 1

    else:
        break

File: scripts/datapull.js
#!/usr/bin/env node

/**
 * Dependencies
 */

var request  = require('superagent');
var mongo = require('mongodb');

/**
 * mongo URI
 */

var mongouri = process.env.MONGOLAB_URI || 'mongodb://localhost/githubdata';

/**
 * Github api URL
 */

var uri = 'https://api.github.com/legacy/repos/search/mongodb?start_page=';

/**
 * Establish mongo connection
 */

var repos, db;

log('starting');
mongo.Db.connect(mongouri, function (err, db_) {
  if (err) throw err;

  db = db_;
  repos = db.collection('repos');
  index(repos);

  log('connected to mongo');

  // start pulling data
  page(0, function (err) {
    log('disconnecting from mongo');

    db.close(function () {
      if (err) {
        log(err);
        return process.exit(1);
      }

      log('completed successfully');
      process.exit(0);
    });
  });
});

/**
 * Fetch a search result from Github by page number
 * and upsert our database
 *
 * This will fire recursively until either Github
 * returns an error or there are no more results.
 *
 * @param {Number} num
 * @param {Function} cb
 */

function page (num, cb) {
  var req = request.get(uri + num);
  req.on('error', cb);
  req.end(function (res) {
    console.log(typeof res, res);
    var pending = res.body && res.body.repositories && res.body.repositories.length;

    if (!pending) return cb();

    // upsert in parallel
    res.body.repositories.forEach(function (repo) {
      fix(repo);

      var match = { username: repo.username, name: repo.name };
      var options = { upsert: true, safe: true };

      repos.update(match, repo, options, function (err, res) {
        if (err) log(err);
        --pending || page(num+1, cb);
      })
    });
  })
}

/**
 * Parse github date strings into real Dates
 *
 * @param {Object} doc
 */

function fix (doc) {
  doc.created = new Date(doc.created);
  doc.created_at = new Date(doc.created_at);
  doc.pushed = new Date(doc.pushed);
  doc.pushed_at = new Date(doc.pushed_at);
}

/**
 * Log the given `msg`
 *
 * @param {String|Error} msg
 */

function log (msg) {
  var prefix = String(new Date) + ': datapull:';
  if (msg instanceof Error)
    console.error(prefix, (msg.stack || '').replace(/\n/g, '\\n'));
  else
    console.log(prefix, msg);
}

/**
 * ensure indexes
 */

function index (col) {
  log('ensuring indexes');
  col.ensureIndex({ username: 1, name: 1 }, { unique: true, safe: true }, function (err) {
    if (err) log(err);
    else log('ensuring indexes complete');
  });
}
